### PR TITLE
Fixed CUDA randint generation for large ranges.

### LINKED ISSUE #125224: Strange behavior of randint using device=cuda
### 🐛 Describe the bug

I want to generate some random numbers on the GPU, but it seems that there is an unreasonably high duplicate rate. 
I can get the expected output under CPU
```python
>>> ls = []
>>> for i in range(2): 
...     ls.extend(torch.randint(0, int(6e9), [int(1e6)], device="cpu").cpu().numpy().tolist())
... 
>>> len(set(ls))
1999678
```
When generating random numbers using torch.randint on GPU, an unexpected behavior is observed.
```python
>>> ls = []
>>> for i in range(2): 
...     ls.extend(torch.randint(0, int(6e9), [int(1e6)], device="cuda:0").cpu().numpy().tolist())
... 
>>> len(set(ls))
1884451
```
Strange behavior of randint using device=cuda, len(set(ls))=1,884,451  means rand produced 2,000,000-1,884,451=115,549 duplicate numbers. This is impossible for the random range of 6e9

I have tried multiple different versions of pytorch on different machines, and the expected value was obtained under pytorch.__version __='1.1.0'(same machines with pytorch.__version __='2.0.1+cu118' failed), but this value is not reasonable in other versions, ranging from 150w to 190w.

Moreover, if the upper bound of randint is changed to 6e8, this issue will not occur, the output 'len(set (ls))' will be reasonable.
```python
>>> ls = []
>>> for i in range(2) :
...     ls.extend(torch.randint(0, int(6e8), [int(1e6)], device="cuda:0").cpu().numpy().tolist())
... 
>>> len(set(ls))
1996659
```

### Versions

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] torch==2.0.1+cu118
[pip3] torchaudio==2.0.1+cu118
[pip3] torchvision==0.15.2+cu118
[pip3] triton==2.0.0
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] torch                     2.0.1+cu118              pypi_0    pypi
[conda] torchaudio                2.0.1+cu118              pypi_0    pypi
[conda] torchvision               0.15.2+cu118             pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

cc @fritzo @neerajprad @alicanb @nikitaved @ptrblck @pbelevich

### PR DESCRIPTION
**Pull request by tringwald**
Fixes #125224

For large ranges, calls to CUDA `randint` use a different `unroll_factor` to generate random ints. This `unroll_factor` was not considered correctly in the calculation of the Philox offsets. Thus, some of the random states were reused, resulting in lower entropy (see #125224).

This also affects multiple other random functions, such as `torch.rand` and `torch.randn`.

### CODE REVIEW

**Comment by tringwald** [comment]
@r-barnes Thanks for reviewing, I added some type annotations and changed the C++ parameters to `const`.

File: test/test_cuda.py [hunk 1]
```diff
@@ -964,6 +964,48 @@ def test_huge_index(self):
...
+            # Measure how many unique numbers are generated in 2 consecutive calls to randint. If random states are
+            # reused, this will yield fewer unique numbers.
+            gen = torch.Generator(device=dev)
+            gen.manual_seed(0)
+            ls = []
+            ls.extend(
+                torch.randint(
+                    0,
+                    high,
+                    torch.Size([1, size]),
```
Comment (eqy) [inline test/test_cuda.py:984]: Out of curiosity, what's the need for generate a `[1, size]` tensor and then immediately `.view(-1)` it? Is it to meet some criteria for the unrolling?

**Review by eqy — APPROVED** [review]

File: test/test_cuda.py (hunk 1 above)
Comment (tringwald) [reply test/test_cuda.py:984]: If I recall correctly, this was necessary to trigger the relevant code path, but I will rewrite this test anyway as I found an easier way to test the fix.

**Comment by eqy** [comment]
CC @drisspg who might know more about the SDPA tests

**Comment by tringwald** [comment]
Thanks @eqy. Those tests in `test_transformers.py` use `torch._fill_mem_eff_dropout_mask_`, which in turn calls a custom CUDA kernel to populate the dropout mask with uniform values before thresholding. I'm not sure why we don't use `torch.rand` there, but it seems like replacing the custom impl with `torch.rand` yields some weird test failures.
I've rolled back the test changes for now, so I can more easily debug the other failures, but we should probably reconsider using a custom rand impl for those tests.

**Comment by tringwald** [comment]
I finally found a way to satisfy all the tests. Do you want to have another look at the changes @eqy? Unfortunately, the `allocator_fuzz` test resulted in an OOM error, so I had to lower the allocation size.

File: aten/src/ATen/native/cuda/DistributionTemplates.h [hunk 2]
```diff
@@ -32,33 +32,31 @@ namespace {
...
   dim3 dim_block(block_size);
   dim3 grid((numel + block_size - 1) / block_size);
   uint32_t blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block_size;
   grid.x = std::min(
       static_cast<uint32_t>(at::cuda::getCurrentDeviceProperties()->multiProcessorCount) * blocks_per_sm,
       grid.x);
   //number of times random will be generated per thread, to offset philox counter in thc random state
-  uint64_t counter_offset = ((numel - 1) / (block_size * grid.x * unroll) + 1)
-                                * curand4_engine_calls;
+  uint64_t counter_offset = ((numel - 1) / (block_size * grid.x * unroll_factor) + 1) * max_generator_offsets_per_curand_call;
```
Comment (eqy) [inline aten/src/ATen/native/cuda/DistributionTemplates.h:59]: Does this `philox_offset` change impact the output of previously correct generation? e.g., after this change would it be expected to get the same RNG output for a deterministic generator via `torch.manual_seed`?

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/cuda/DistributionTemplates.h (hunk 2 above)
Comment (tringwald) [reply aten/src/ATen/native/cuda/DistributionTemplates.h:59]: It should only affect the previously wrong generations. I superficially checked my build vs. `torch=2.3.0` and some differences are (for CUDA only):
-  `torch.rand/torch.randn` calls with `dtype=torch.double`
-  `torch.randint()` calls for `dtype=int64, double, float, bfloat16` when `to-from` range is above 2**32 (`float` and `bfloat16` would have precision limitations in that range anyway). 
- `RReLU` with `dtype=torch.double`

Basically every call of `distribution_nullary_kernel` with `curand4_engine_calls/2` will have different results now. Should we add a `UserWarning` in those cases?

**Comment by tringwald** [comment]
Can you have a look at this? @malfet 
The changes are BC-breaking for certain random function calls, i.e. the sequence of generated numbers will be different.Should we add a `UserWarning` for this?

**Comment by tringwald** [comment]
I need additional approval for this PR. Could either of you take a look at this? @lezcano @Skylion007

**Review by lezcano — APPROVED** [review]
Approving but didn't review. @eqy did.
